packages/notion-dump/src/lib/domloadImage.ts

```diff
@@ -0,0 +1,114 @@
+import * as fs from "fs";
+import * as path from "path";
+
+async function downloadImage({
+  url,
+  outputPath,
+}: {
+  url: string;
+  outputPath: string;
+}): Promise<string> {
+  const response = await fetch(url);
+  const arrayBuffer = await response.arrayBuffer();
+  await fs.promises.writeFile(outputPath, Buffer.from(arrayBuffer));
+
+  // Return the Content-Type header
+  return response.headers.get("Content-Type") || "";
+}
+
+function getFileExtension(contentType: string, originalUrl: string): string {
+  // First, try to determine extension from Content-Type
+  const mimeToExt: { [key: string]: string } = {
+    "image/jpeg": ".jpg",
+    "image/png": ".png",
+    "image/gif": ".gif",
+    "image/webp": ".webp",
+    "image/svg+xml": ".svg",
+  };
+
+  const ext = mimeToExt[contentType];
+  if (ext) return ext;
+
+  // If Content-Type doesn't help, try to extract from the URL
+  const urlParts = originalUrl.split("/");
+  const filename = urlParts[urlParts.length - 1].split("?")[0];
+  const fileExtension = path.extname(filename);
+  if (fileExtension) return fileExtension;
+
+  // Default to .jpg if we can't determine the extension
+  return ".jpg";
+}
```

Comment:
p4; It’s already great, but I think it could be even better with the following changes:
  
- Defined SupportedImageMimeType and SupportedImageExtension types to enhance type safety.
- Separated DEFAULT_IMAGE_EXTENSION as a constant to explicitly express the default value.
- Moved mimeTypeToExtensionMap outside of the function and used the Record type to improve type safety.
- Separated getFileExtensionFromContentType and getFileExtensionFromUrl functions to clarify their roles.
- Renamed variables for better clarity, such as changing mimeToExt to mimeTypeToExtensionMap.
- Added type checking within the getFileExtension function to ensure that the extracted extension from the URL is a supported extension.

These changes improve the code’s type safety and clarify the responsibilities of each part. Additionally, separating constants and maps outside the functions enhances code reusability and maintainability.

```suggestion
import * as path from "path";

type SupportedImageMimeType = 
  | "image/jpeg"
  | "image/png"
  | "image/gif"
  | "image/webp"
  | "image/svg+xml";

type SupportedImageExtension = ".jpg" | ".png" | ".gif" | ".webp" | ".svg";

const DEFAULT_IMAGE_EXTENSION: SupportedImageExtension = ".jpg";

const mimeTypeToExtensionMap: Record<SupportedImageMimeType, SupportedImageExtension> = {
  "image/jpeg": ".jpg",
  "image/png": ".png",
  "image/gif": ".gif",
  "image/webp": ".webp",
  "image/svg+xml": ".svg",
};

function getFileExtensionFromContentType(contentType: string): SupportedImageExtension | undefined {
  return mimeTypeToExtensionMap[contentType as SupportedImageMimeType];
}

function getFileExtensionFromUrl(url: string): string {
  const urlSegments = url.split("/");
  const filenameWithQuery = urlSegments[urlSegments.length - 1];
  const filename = filenameWithQuery.split("?")[0];
  return path.extname(filename);
}

function getFileExtension(contentType: string, originalUrl: string): SupportedImageExtension {
  // First, try to determine extension from Content-Type
  const extensionFromContentType = getFileExtensionFromContentType(contentType);
  if (extensionFromContentType) return extensionFromContentType;

  // If Content-Type doesn't help, try to extract from the URL
  const extensionFromUrl = getFileExtensionFromUrl(originalUrl);
  if (extensionFromUrl && Object.values(mimeTypeToExtensionMap).includes(extensionFromUrl as SupportedImageExtension)) {
    return extensionFromUrl as SupportedImageExtension;
  }

  // Default to .jpg if we can't determine the extension
  return DEFAULT_IMAGE_EXTENSION;
}
```

Reply:
p3; 
When written as `parameterized tests`(e.g. `each.test()`), test code enhances code safety and serves as documentation and case studies. Such test cases can be helpful when extending image types, especially for functions like these.


import { describe, it, expect } from 'vitest';
import { 
  getFileExtensionFromContentType, 
  getFileExtensionFromUrl, 
  getFileExtension,
  SupportedImageExtension,
  DEFAULT_IMAGE_EXTENSION
} from './fileExtensionUtils';
```ts
describe('File Extension Utilities', () => {
  describe('getFileExtensionFromContentType', () => {
    it.each([
      ['image/jpeg', '.jpg'],
      ['image/png', '.png'],
      ['image/gif', '.gif'],
      ['image/webp', '.webp'],
      ['image/svg+xml', '.svg'],
    ])('should return %s for MIME type %s', (mimeType, expectedExtension) => {
      expect(getFileExtensionFromContentType(mimeType)).toBe(expectedExtension);
    });

    it.each([
      ['image/tiff'],
      ['application/pdf'],
    ])('should return undefined for unsupported MIME type %s', (mimeType) => {
      expect(getFileExtensionFromContentType(mimeType)).toBeUndefined();
    });
  });

  describe('getFileExtensionFromUrl', () => {
    it.each([
      ['https://example.com/image.jpg', '.jpg'],
      ['https://example.com/path/to/image.png?param=value', '.png'],
      ['https://example.com/image', ''],
      ['https://example.com/path/to/image?param=value', ''],
    ])('should extract correct extension "%s" from URL %s', (url, expectedExtension) => {
      expect(getFileExtensionFromUrl(url)).toBe(expectedExtension);
    });
  });

  describe('getFileExtension', () => {
    it.each([
      ['image/png', 'https://example.com/image.jpg', '.png'],
      ['image/tiff', 'https://example.com/image.gif', '.gif'],
      ['image/tiff', 'https://example.com/image', DEFAULT_IMAGE_EXTENSION],
      ['image/png', 'https://example.com/image.jpg', '.png'],
      ['image/tiff', 'https://example.com/image.bmp', DEFAULT_IMAGE_EXTENSION],
    ])('should return %s for content type %s and URL %s', (contentType, url, expectedExtension) => {
      expect(getFileExtension(contentType, url)).toBe(expectedExtension);
    });
  });
});
```

Reply:
thank you for review. I write this test code